doc/md/15-motoko-style.md

```diff
@@ -837,62 +861,523 @@ Rationale: `g[1]` in particular will be misparsed as an indexing operation.
+- Use `await` judiciously.
+  - Minimize the number of `await` calls in a function to avoid unnecessary suspensions.
 
--   Limit each module to a single main concept/type/class or closely entangled family of concepts/types/classes.
+    ```motoko no-repl
+    let user = await db.getUser(id); // Good ~4s
+    let orders = await db.getOrders(user); // Bad: sequential calls introduce high latency ~6s
+    ```
+
+- Use `async*` and `await*` for same-canister calls.
+  - When calling an async function **within the same canister**, use `async*` to declare the function and `await*` to call it. This avoids unnecessary inter-canister message overhead.
+
+  ```motoko no-repl
+  func internalLogic(): async* () {
+    //Some internal logic
+  };
+
+  public func intraCall(): async (){
+    await* internalLogic()
+  }
+  ```
+
+- Use composite queries where applicable.
+
+Example:
+
+```motoko no-repl
+// This is an unecessary update call :
+public func couldBeQuery() : async () {
+  await queryFunction();
+}
+```
+
+This can be optimized with a composite query:
+
+```motoko no-repl
+// Composite queries can await other queries without becoming an update call
+public composite query func betterQuery() : async () {
+  await queryFunction();
+}
+```
+
+:::note
+Composite queries cannot perform or persist state changes. Any modifications made during a composite query will be rolled back automatically. Use them only when no mutation is required.
+:::
+
+### Pattern matching
+
+- Prefer `let-else` when one path returns or traps.
+  When destructuring a `Result` or `Option` where the failure case immediately exits the function (e.g. via `return`, or `trap`), use `let-else` to avoid deep nesting and improve readability.
+
+  ```motoko no-repl
+  // Extract a value from an Option type
+  func extractUserName(userOpt: ?Text) : Text {
+    // Without let-else (verbose)
+    // switch (userOpt) {
+    //   case (null) { return "Guest"; };
+    //   case (?name) { return "User: " # name; };
+    // };
+    // With let-else (concise)
+    let ?name = userOpt else return "Guest";
+    return "User: " # name;
+  }
+  ```
+
+- Use `switch` instead of manual conditionals for variants.
+  - Prefer exhaustive pattern matching to avoid missing cases.
+  - Exhaust all cases before using the wildcard pattern.
+
+    ```motoko no-repl
+    switch (response) {
+      case (#possiblity1) { processPossibility1() };
+      case (#possiblity2) { processPossibility2() }
+      case (#possiblity3) { processPossibility3() };
+      case (_) { handleExecption(err) };;
+    };
+    ```
+
+### Data structures
+
+- Prefer `List` over dynamic array resizing.
+  - `Array` in Motoko is immutable, so if dynamic resizing is needed, `List` is a better choice.
+
+    ```motoko no-repl
+    let list = List.empty<Nat>()
+    List.add(list, 1);;
+    let arr = List.toArray(list);
+    ```
 
-<!--
-=== To be extended
--->
+## Motoko grammar
```

Comment:
I wouldn't include the grammar but link to it instead, since it is often updated as the language changes. It used to live in doc/md/examples/grammar.txt or similar.